coll/gpu: Stream read bcast and alltoall

dycz0fx opened the pull request:
## Pull Request Description
Add an intra-node MPI_Bcast and MPI_Alltoall algorithm (stream_read) with Level 0 API for Intel GPU. 
In MPI_Bcast, non-root processes read data from the root's send buffer, and in MPI_Alltoall, every process reads data from other processes' send buffers.

Last 6 commits only
Depends on ~~#6571~~ #6587

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

dycz0fx commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/ch4/gpu

dycz0fx commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/ch4/gpu

dycz0fx commented:
This PR is ready for review.

hzhou commented on src/mpid/ch4/shm/posix/posix_coll_gpu_streamer.h:
```diff
@@ -0,0 +1,134 @@
+    }
+    if (errs > 0) {
+        goto fallback;
+    }
```
Every process allocate ipc_handle and use `Allgather` only for this error checking, right? If we add device->host version of local copy, then we only need `Bcast` the root ipc_handle, right?

dycz0fx replied:
Yes, this Allgather is for the error checking. It adds some overhead compared to Bcast, but we need to make sure all the buffers are on GPU. Also, this algorithm is targeted for large messages, so the added overhead does not affect the overall performance too much. Only messages larger than this CVAR ( MPIR_CVAR_BCAST_STREAM_READ_MSG_SIZE_THRESHOLD) will use the stream read bcast. 

In this algorithm, it is kind of important to keep the root's sendbuf on the GPU so when other ranks copy the send_buf to its own GPU buffer, the communications go through separate Xelinks directly to maximize the bandwidth.

hzhou reviewed (commented):
Look good. Just question on the bcast algorithm. What the algorithm is called stream_read? I wouldn't able to guess any clue from "stream_read". How about just name it `ipc` or `ipc_gpu`? The commit message says it is for Intel GPU, but the code seems to be generic. Is CUDA prevented to hit this path?

dycz0fx commented:
> Look good. Just question on the bcast algorithm. What the algorithm is called stream_read? I wouldn't able to guess any clue from "stream_read". How about just name it `ipc` or `ipc_gpu`? The commit message says it is for Intel GPU, but the code seems to be generic. Is CUDA prevented to hit this path?

Right, maybe "ipc" is a better name for it now. Initially, these algorithms are implemented with low-level level zero API. The stream refers to these algorithms copy the data using streamers instead of kernels. Now we use the MPIR_Localcopy_gpu and MPL_gpu_imemcpy to move the data with IPC handles, so using "ipc" would make more sense.

I think this code would work on Nvidia GPUs, but I am not sure about the performance as it depends on the P2P implementation. Since we have never tested the code with Nvidia GPU, I will try to prevent CUDA to hit this path just to be safe.

hzhou commented:
> I think this code would work on Nvidia GPUs, but I am not sure about the performance as it depends on the P2P implementation. Since we have never tested the code with Nvidia GPU, I will try to prevent CUDA to hit this path just to be safe.

To the opposite, let's hit the CUDA path and test see if it works :)  Could you add a test for both algorithm?

dycz0fx commented:
> > I think this code would work on Nvidia GPUs, but I am not sure about the performance as it depends on the P2P implementation. Since we have never tested the code with Nvidia GPU, I will try to prevent CUDA to hit this path just to be safe.
> 
> To the opposite, let's hit the CUDA path and test see if it works :) Could you add a test for both algorithm?

I see. Sure, I can add a test.

dycz0fx commented:
test:mpich/ch4/gpu

dycz0fx commented:
test:mpich/ch4/gpu

dycz0fx commented:
- Renamed the algorithm name from stream read to ipc read. 
- Added two tests (alltoall_gpu.c and bcast_gpu.c in test/mpi/coll) to test the ipc read bcast and alltoall. These tests perform bcast and alltoall on GPU with a contiguous datatype.
Currently, these two algorithms do not work for CUDA.
    - Alltoall: This algorithm uses MPL_gpu_imemcpy to move the data, and this function is not implemented in CUDA.
    - Bcast: This algorithm uses MPIR_Localcopy_gpu to move the data. With Intel GPU, MPIR_Localcopy_gpu invokes MPL_gpu_imemcpy and MPL_gpu_fast_memcpy (both not implemented in CUDA). With Nvidia GPU, MPIR_Localcopy_gpu fallbacks to do_localcopy, and the receive buffer seems to be incorrect in this case ([https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ofi/211/compiler=gnu,jenkins_c[…]ILITY_AFFINITY_CONTIGUOUS_DEVICE_MPIR_CVAR_ENABLE_GPU_1/](https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ofi/211/compiler=gnu,jenkins_configure=debug,label=gpu,netmod=ofi/testReport/junit/(root)/coll/00273_____coll_bcast_gpu_6__memtype_all_MPIR_CVAR_BCAST_COMPOSITION_1_MPIR_CVAR_CH4_IPC_GPU_P2P_THRESHOLD_1_MPIR_CVAR_BCAST_POSIX_INTRA_ALGORITHM_ipc_read_MPIR_CVAR_BCAST_IPC_READ_MSG_SIZE_THRESHOLD_1024_MTEST_GPU_VISIBILITY_AFFINITY_CONTIGUOUS_DEVICE_MPIR_CVAR_ENABLE_GPU_1/)).

hzhou commented:
> * Renamed the algorithm name from stream read to ipc read.
> 
>     * Added two tests (alltoall_gpu.c and bcast_gpu.c in test/mpi/coll) to test the ipc read bcast and alltoall. These tests perform bcast and alltoall on GPU with a contiguous datatype.
>       Currently, these two algorithms do not work for CUDA.
>       
>       * Alltoall: This algorithm uses MPL_gpu_imemcpy to move the data, and this function is not implemented in CUDA.
>       * Bcast: This algorithm uses MPIR_Localcopy_gpu to move the data. With Intel GPU, MPIR_Localcopy_gpu invokes MPL_gpu_imemcpy and MPL_gpu_fast_memcpy (both not implemented in CUDA). With Nvidia GPU, MPIR_Localcopy_gpu fallbacks to do_localcopy, and the receive buffer seems to be incorrect in this case ([https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ofi/211/compiler=gnu,jenkins_c[…]ILITY_AFFINITY_CONTIGUOUS_DEVICE_MPIR_CVAR_ENABLE_GPU_1/](https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ofi/211/compiler=gnu,jenkins_configure=debug,label=gpu,netmod=ofi/testReport/junit/(root)/coll/00273_____coll_bcast_gpu_6__memtype_all_MPIR_CVAR_BCAST_COMPOSITION_1_MPIR_CVAR_CH4_IPC_GPU_P2P_THRESHOLD_1_MPIR_CVAR_BCAST_POSIX_INTRA_ALGORITHM_ipc_read_MPIR_CVAR_BCAST_IPC_READ_MSG_SIZE_THRESHOLD_1024_MTEST_GPU_VISIBILITY_AFFINITY_CONTIGUOUS_DEVICE_MPIR_CVAR_ENABLE_GPU_1/)).

Thanks for the testing! Lets xfail the tests for now. Please open an issue for `MPL_gpu_imemcpy` not working in CUDA and link the issue in the xfail entries.

dycz0fx commented:
> > * Renamed the algorithm name from stream read to ipc read.
> >   
> >   * Added two tests (alltoall_gpu.c and bcast_gpu.c in test/mpi/coll) to test the ipc read bcast and alltoall. These tests perform bcast and alltoall on GPU with a contiguous datatype.
> >     Currently, these two algorithms do not work for CUDA.
> >     
> >     * Alltoall: This algorithm uses MPL_gpu_imemcpy to move the data, and this function is not implemented in CUDA.
> >     * Bcast: This algorithm uses MPIR_Localcopy_gpu to move the data. With Intel GPU, MPIR_Localcopy_gpu invokes MPL_gpu_imemcpy and MPL_gpu_fast_memcpy (both not implemented in CUDA). With Nvidia GPU, MPIR_Localcopy_gpu fallbacks to do_localcopy, and the receive buffer seems to be incorrect in this case ([https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ofi/211/compiler=gnu,jenkins_c[…]ILITY_AFFINITY_CONTIGUOUS_DEVICE_MPIR_CVAR_ENABLE_GPU_1/](https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-gpu-ofi/211/compiler=gnu,jenkins_configure=debug,label=gpu,netmod=ofi/testReport/junit/(root)/coll/00273_____coll_bcast_gpu_6__memtype_all_MPIR_CVAR_BCAST_COMPOSITION_1_MPIR_CVAR_CH4_IPC_GPU_P2P_THRESHOLD_1_MPIR_CVAR_BCAST_POSIX_INTRA_ALGORITHM_ipc_read_MPIR_CVAR_BCAST_IPC_READ_MSG_SIZE_THRESHOLD_1024_MTEST_GPU_VISIBILITY_AFFINITY_CONTIGUOUS_DEVICE_MPIR_CVAR_ENABLE_GPU_1/)).
> 
> Thanks for the testing! Lets xfail the tests for now. Please open an issue for `MPL_gpu_imemcpy` not working in CUDA and link the issue in the xfail entries.

Got it. Will do that.

dycz0fx commented:
test:mpich/ch4/gpu

hzhou reviewed (approved):
LGTM
